### PR TITLE
Matrix derivatives. Examples from the Cookbook

### LINKED ISSUE #5858: Implement the formulas from the matrix cookbook
See http://orion.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf .  These should all be implemented, e.g., with MatrixExpr.  In particular, currently differentiation with respect to matrices is wrong because it doesn't handle transposes correctly.  For example, compare the output of this to that in the cookbook:

``` python
In [88]: A = MatrixSymbol("A", n, n)

In [89]: x = MatrixSymbol("x", n, 1)

In [90]: diff(A.T*x, x)
Out[90]: A'
```

According to equation 61 in the cookbook, this should just be A.

Original issue for #5858: http://code.google.com/p/sympy/issues/detail?id=2759
Original author: https://code.google.com/u/asmeurer@gmail.com/
Original owner: https://code.google.com/u/asmeurer@gmail.com/

### PR DESCRIPTION
**Pull request by Upabjojr**
Partly fixes https://github.com/sympy/sympy/issues/5858

### CODE REVIEW

**Comment by Upabjojr** [comment]
Hi @jacknlliu , I've seen you were asking about matrix differentiation. Maybe you are interested in this PR.

File: sympy/matrices/expressions/tests/test_derivatives.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    assert expr._eval_derivative(x) == B*x + B.T*x
+
+    # Cookbook example 82:
+    expr = b.T*X.T*D*X*c
+    assert expr._eval_derivative(X) == D.T*X*b*c.T + D*X*c*b.T
+
+    # Cookbook example 83:
+    expr = (X*b + c).T*D*(X*b + c)
+    # TODO: parsing error
+    # assert expr._eval_derivative(X) == D*X*b*b.T + D*c*b.T + D.T*X*b*b.T + D.T*c*b.T
```
Comment (Upabjojr) [inline sympy/matrices/expressions/tests/test_derivatives.py]: @asmeurer this is the only failed example from the cookbook so far, parsing back from the index-summation is currently failing.

**Comment by Upabjojr** [comment]
I have skipped the examples with the determinant, as I haven't implemented the handling of the determinant yet.

**Comment by Upabjojr** [comment]
@karlnapf this could be of interest to you.

File: sympy/matrices/expressions/tests/test_derivatives.py (hunk 1 above)
Comment (asmeurer) [reply sympy/matrices/expressions/tests/test_derivatives.py]: What happens when it fails? Do you get an exception or an invalid expression or what?

Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]: This case raises an exception, an _IndexError_ in one of the parsing arrays. This is definitely a bug.

**Comment by Upabjojr** [comment]
@asmeurer I think I have fixed the previous issue. We'd need a way to represent tensor products of matrices to be able to represent all kinds of derivatives, as the matrix derivative increases the tensor-rank of the expression.

If there are no objections, I'm going to merge this PR.

**Comment by maaaaadn** [comment]
[CalculateGradientsOfProjection.ipynb.zip](https://github.com/sympy/sympy/files/1521791/CalculateGradientsOfProjection.ipynb.zip)


I'm facing a problem with the derivative. 
The error reads 

> TypeError: _entry() got an unexpected keyword argument 'expand'

since I'm quite fresh to sympy and python in general I'm not familiar how to solve this issue. 
I've attached a pdf of the problem plus the .ipynb jupyter project (in a .zip). 
[CalculateGradientsOfProjection.pdf](https://github.com/sympy/sympy/files/1521789/CalculateGradientsOfProjection.pdf)

thanks in advance!

**Pull request merged**